### PR TITLE
coord: Guard against race condition between Coordinator and SessionClient

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR should guard against a race condition where the `Coordinator` responds to a request successfully, but a `SessionClient` never observes the response and thus never gets back the `Session` for termination, effectively leaking the `Session`.

~~I haven't tried reproducing this race, but I think if we put a sleep [here](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/client.rs#L818) on `main`, and then close the resulting connection, we should see the Session get leaked.~~

Was able to repro the original issue with Nikhil's test script after running for ~1 minute. After running the same script for ~10 minutes with this PR I could not reproduce the issue.

```
#!/usr/bin/env bash

while :; do
  timeout=$(python3 -c 'import random; print(random.uniform(0.01, 1.0))')
  rows=$(python3 -c 'import random; print(random.uniform(1, 100000))')
  echo "timeout: $timeout"
  echo "rows: $rows"
  curl localhost:6876/api/sql --data "{\"query\": \"select generate_series(1, $rows::int8)\"}" -H 'content-type: application/json' -m "$timeout" &
done
```

### Motivation

Possible solution for https://github.com/MaterializeInc/materialize/issues/26154

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/adapter/src/client.rs [hunk 1]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
         msg
     }
 }
+
+/// ~~SPOOKY ZONE~~
+///
+/// There is a small race that can occur with [`tokio`], a [`SessionClient`], and the `Coordinator`
```
Comment (guswynn) [inline src/adapter/src/client.rs:1045]:
```suggestion
/// A naive handling of a [`SessionClient`] interacting with the semantics of a [`tokio::sync::oneshot`] can cause a race, that this drop guard works around.
```

File: src/adapter/src/client.rs [hunk 2]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
...
+/// that requires this Drop guard.
+///
+/// 1. A [`SessionClient`] can execute a command, sending it's owned [`Session`] to the
+///    Coordinator.
+/// 2. While processing this request, the user closes their connection to Materialize.
+/// 3. The Coordinator (which is run concurrently on a different thread and in a different task)
```
Comment (guswynn) [inline src/adapter/src/client.rs:1051]:
```suggestion
/// 3. The Coordinator (which is run concurrently in a different task executing on its own thread).
```

File: src/adapter/src/client.rs [hunk 3]
```diff
@@ -811,30 +811,37 @@ impl SessionClient {
...
                             .metrics
                             .commands
                             .with_label_values(&[typ, status, name_hint.as_str()])
                             .inc();
                     }
-                    self.session = Some(res.session);
+                    guard.client.session = Some(res.session);
                     return res.result
                 },
                 _err = &mut cancel_future, if !cancelled => {
```
Comment (guswynn) [inline src/adapter/src/client.rs:842]: While we are here, we should add docs saying that the `cancel_safe` future can be cancelled partially completed (actually, is this another bug?)

File: src/adapter/src/client.rs [hunk 4]
```diff
@@ -811,30 +811,37 @@ impl SessionClient {
             cmd
         });
 
+        // SPOOKY ZONE! See the doc comment on `SessionTerminateGuard` for why this exists.
+        //
+        // tl;dr this guard prevents a race where a `Session` is returned on `rx` but never placed
+        // back in `self` because the Future returned by this function is concurrently dropped.
```
Comment (guswynn) [inline src/adapter/src/client.rs:817]: can this doc explain where the `send_with_cancel` future is dropped? Do we abort tasks with closed connections, or does hyper?

File: src/adapter/src/client.rs [hunk 5]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
         msg
     }
 }
+
+/// ~~SPOOKY ZONE~~
+///
+/// There is a small race that can occur with [`tokio`], a [`SessionClient`], and the `Coordinator`
+/// that requires this Drop guard.
+///
+/// 1. A [`SessionClient`] can execute a command, sending it's owned [`Session`] to the
```
Comment (benesch) [inline src/adapter/src/client.rs:1048]:
```suggestion
/// 1. A [`SessionClient`] can execute a command, sending its owned [`Session`] to the
```

File: src/adapter/src/client.rs [hunk 6]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
...
+impl<'a, T> Drop for SessionTerminateGuard<'a, T> {
+    fn drop(&mut self) {
+        let SessionTerminateGuard { client, rx } = self;
+
+        // We need to explicitly close the channel to make sure if the Coordinator tries sending
+        // a response, it will fail. If the sending fails the Coordinator will then terminate the
+        // session for us.
+        rx.close();
+
+        // Now that the channel is guaranteed close, lets check if we ever received a message and
```
Comment (benesch) [inline src/adapter/src/client.rs:1094]:
```suggestion
        // Now that the channel is guaranteed closed, let's check if we ever received a message and
```

File: src/adapter/src/client.rs [hunk 7]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
...
+///                 dropped mid-processing we can check for this race and re-assign the [`Session`]
+///                 to the [`SessionClient`] so it can properly terminate the session as part of
+///                 its Drop impl.
+pub struct SessionTerminateGuard<'a, T> {
+    client: &'a mut SessionClient,
+    rx: oneshot::Receiver<Response<T>>,
+}
+
+impl<'a, T> Future for SessionTerminateGuard<'a, T> {
+    type Output = Result<Response<T>, tokio::sync::oneshot::error::RecvError>;
```
Comment (benesch) [inline src/adapter/src/client.rs:1075]:
```suggestion
    type Output = Result<Response<T>, oneshot::error::RecvError>;
```

File: src/adapter/src/client.rs [hunk 8]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
...
+
+    fn poll(
+        mut self: pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
```
Comment (benesch) [inline src/adapter/src/client.rs:1080]: Can you `use std::task::{Context, Poll}` and `use std::pin::Pin` here?

File: src/adapter/src/client.rs [hunk 9]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
...
+/// There is a small race that can occur with [`tokio`], a [`SessionClient`], and the `Coordinator`
+/// that requires this Drop guard.
+///
+/// 1. A [`SessionClient`] can execute a command, sending it's owned [`Session`] to the
+///    Coordinator.
+/// 2. While processing this request, the user closes their connection to Materialize.
+/// 3. The Coordinator (which is run concurrently on a different thread and in a different task)
+///    finishes processing our request and is able to successfully send a response over the
+///    provided [`oneshot::Sender`].
+/// 4. [`tokio`] realizes the connection is closed and Drops the task which is awaiting the result
```
Comment (guswynn) [inline src/adapter/src/client.rs:1054]: Why can the `tx` send the command (with the session inside) without an error? Is it because the receiver hasn't quite been dropped yet (Or is in the process of being dropped?)

File: src/adapter/src/client.rs [hunk 10]
```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
...
+}
+
+impl<'a, T> Future for SessionTerminateGuard<'a, T> {
+    type Output = Result<Response<T>, tokio::sync::oneshot::error::RecvError>;
+
+    fn poll(
+        mut self: pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        self.rx.poll_unpin(cx)
```
Comment (benesch) [inline src/adapter/src/client.rs:1081]: In terms of separation of responsibilities, can we put this guard entirely in charge of restoring the `Session`? It seems like this future could always handle the `Response` decomposition and session restoration, and then hand back an `Output` of `Result<T, AdapterError>`.

**Review by guswynn — COMMENTED** [review]
This looks good, but I have some questions I want to clarify first.

Also, I am thinking if there is a way to model this to avoid issues like this; for example, how do tasks running http responses get aborted? I don't remember tokio or hyper doing this, so I'd love to learn more about that

File: src/adapter/src/client.rs (hunk 3 above)
Comment (benesch) [reply src/adapter/src/client.rs:842]: Sure, probably worth documenting on `execute` that `cancel_future` must be cancel safe. It's not a bug AFAICT because the one future that we pass in here is in fact cancel safe.

File: src/adapter/src/client.rs (hunk 9 above)
Comment (benesch) [reply src/adapter/src/client.rs:1054]: Yeah, exactly. The `Drop` impl of `SessionClient` runs before the `Drop` impl of `Receiver` runs and closes the channel. And in general the only way to close a oneshot channel without losing the message inside is to call `close()` on it manually.

Comment (guswynn) [reply src/adapter/src/client.rs:1054]: Yeah, til about `close()`+`try_recv()`, which is quite nice!

**Review by mjibson — COMMENTED** [review]
Looks reasonable to me! Leaving signoff to other reviewers.

File: src/adapter/src/client.rs (hunk 5 above)
Comment (ParkMyCar) [reply src/adapter/src/client.rs]: Thanks!

File: src/adapter/src/client.rs (hunk 8 above)
Comment (ParkMyCar) [reply src/adapter/src/client.rs]: Done!

File: src/adapter/src/client.rs (hunk 10 above)
Comment (ParkMyCar) [reply src/adapter/src/client.rs]: Good idea, refactored and made this change!

File: src/adapter/src/client.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/adapter/src/client.rs]: Done!

File: src/adapter/src/client.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/adapter/src/client.rs:851]: Added!

**Comment by ParkMyCar** [comment]
> Also, I am thinking if there is a way to model this to avoid issues like this; for example, how do tasks running http responses get aborted? I don't remember tokio or hyper doing this, so I'd love to learn more about that

I'm not sure of the exact behavior, I asked in [Discord](https://discord.com/channels/500028886025895936/670880858630258689/1220381786601685032). Thinking about it a bit more though, I wonder if this is only a problem for the HTTP endpoints because `axum` is what spawns and would thus abort a tokio task?

**Comment by def-** [comment]
@ParkMyCar  Locally I'm still seeing this fail:
```
bin/mzcompose --find cluster down && bin/mzcompose --find cluster run test-http-race-condition
```
If I remove the http calls in the test it succeeds.

**Comment by ParkMyCar** [comment]
> I added one fixed test in cluster and one randomized in parallel-workload.
> 
> @ParkMyCar Locally I'm still seeing this fail:
> 
> ```
> $ bin/mzcompose --find cluster down && bin/mzcompose --find cluster run test-http-race-condition
> [...]
> AssertionError: There are supposed to be no sessions remaining, but there are:
> (['29939', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 17, 627000, tzinfo=datetime.timezone.utc)], ['513794', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 26, 806000, tzinfo=datetime.timezone.utc)], ['492841', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 8, 154000, tzinfo=datetime.timezone.utc)], ['140605', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 27, 4000, tzinfo=datetime.timezone.utc)], ['480061', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 32, 211000, tzinfo=datetime.timezone.utc)], ['363596', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 33, 615000, tzinfo=datetime.timezone.utc)], ['459854', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 22, 962000, tzinfo=datetime.timezone.utc)], ['87397', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 57, 1000, tzinfo=datetime.timezone.utc)], ['206182', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 42, 781000, tzinfo=datetime.timezone.utc)], ['125861', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 54, 497000, tzinfo=datetime.timezone.utc)], ['455856', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 23, 203000, tzinfo=datetime.timezone.utc)], ['203185', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 11, 967000, tzinfo=datetime.timezone.utc)], ['77777', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 37, 494000, tzinfo=datetime.timezone.utc)], ['403701', 'u2', datetime.datetime(2024, 3, 21, 15, 33, 15, 671000, tzinfo=datetime.timezone.utc)], ['281086', 'u2', datetime.datetime(2024, 3, 21, 15, 32, 56, 120000, tzinfo=datetime.timezone.utc)])
> ```
> 
> If I remove the http calls in the test it succeeds.

Thanks for adding the test, let me run this locally and debug!

**Comment by def-** [comment]
The parallel-workload one fails all over too: https://buildkite.com/materialize/nightlies/builds/6996 But hopefully that's the same issue and will get fixed too.

File: src/adapter/src/client.rs [hunk 11]
```diff
@@ -166,10 +166,23 @@ impl Client {
         let notice_tx = session.retain_notice_transmitter();
 
         let (tx, rx) = oneshot::channel();
+
+        // ~~SPOOKY ZONE~~
```
Comment (def-) [inline src/adapter/src/client.rs:170]: 👻

**Review by def- — APPROVED** [review]

**Review by benesch — APPROVED** [review]
Code LGTM. Pushed up some small changes to the tests (see add'l two commits) to 1) ensure that the parallel workload test closes connections before asserting that `mz_sessions` is empty and 2) only filter out the current connection, rather than doing any reasoning based on connection time.

**Comment by def-** [comment]
I rebased to resolve conflicts and get tests pipeline green (Thank you Paraguay's DST).

Thanks, the test changes look fine, assuming https://buildkite.com/materialize/nightlies/builds/7040 is green now.

**Comment by benesch** [comment]
Pushed up a fix for the Python formatting error.
